Userland/Libraries/LibWeb/DOM/Node.cpp

```diff
@@ -996,16 +996,44 @@ u16 Node::compare_document_position(JS::GCPtr<Node> other)
     if ((node1 == nullptr || node2 == nullptr) || (&node1->root() != &node2->root()))
         return DOCUMENT_POSITION_DISCONNECTED | DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC | (node1 > node2 ? DOCUMENT_POSITION_PRECEDING : DOCUMENT_POSITION_FOLLOWING);
 
+    Vector<Node*> node1_ancestors;
+    for (auto* node = node1; node; node = node->parent())
+        node1_ancestors.append(node);
+
+    Vector<Node*> node2_ancestors;
+    for (auto* node = node2; node; node = node->parent())
+        node2_ancestors.append(node);
+
+    auto it_node1_ancestors = node1_ancestors.rbegin();
+    auto it_node2_ancestors = node2_ancestors.rbegin();
+    // Walk ancestor chains of both nodes starting from root
+    while (it_node1_ancestors != node1_ancestors.rend() && it_node2_ancestors != node2_ancestors.rend()) {
+        auto* ancestor1 = *it_node1_ancestors;
+        auto* ancestor2 = *it_node2_ancestors;
+
+        // If ancestors of nodes at the same level in the tree are different then preceding node is the one with lower sibling position
+        if (ancestor1 != ancestor2) {
+            return ancestor1->index() > ancestor2->index() ? DOCUMENT_POSITION_FOLLOWING : DOCUMENT_POSITION_PRECEDING;
+        }
```

Comment:
What happens if one or both nodes are `Attr` nodes?
They don't participate in the regular node sibling chain, so they will have overlapping indexes.

Reply:
If you want to do even less work here, you can follow the sibling chain from `ancestor1` to `ancestor2`. If you reach the end, `ancestor1` is after `ancestor2`. That way you don't need to walk the siblings twice to compute their index :)

Reply:
> If you want to do even less work here, you can follow the sibling chain from ancestor1 to ancestor2.

makes sense. I rewrote it that way. thank you.

> What happens if one or both nodes are Attr nodes?
They don't participate in the regular node sibling chain, so they will have overlapping indexes.

for some reason algorithm described in the spec takes in account attrs only when one node is ancestor of another (steps 7, 8).